if function is not in the function map, treat as if it has no body

polgreen opened the pull request:
For the call graph, we add the function to the call graph but do not try to look for function calls in the body.
For slicing global inits, we do not try to look in the function body for symbols.
This fixes issue https://github.com/diffblue/cbmc/issues/2631

polgreen committed:
if function is not in the function map, treat as if it has no body

For the call graph, we add the function to the call graph but do not try to look for function calls in the body.
For slicing global inits, we do not try to look in the function body for symbols.
This fixes issue https://github.com/diffblue/cbmc/issues/2631

polgreen commented:
I can't access the AWS code build, my AWS account says "No build found
Unable to find build cbmc-windows:9c0f71ed-92dc-4ed1-8ed6-29254ccf3638". Could someone please send me the output?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ed33edead36efbe5eaa9dfb6af885b0ae9587a5b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80457679

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9e8d97388e0a59de18174432f1135ab050303aa3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80462217

smowton commented on regression/goto-instrument/assembly_call_graph_test/main.c:
```diff
@@ -0,0 +1,37 @@
+#define assert(e) __CPROVER_assert(e,"error")
+#endif
+
+volatile int turn; // integer variable to hold the ID of the thread whose turn is it
```
Can this test case be minimised to do just enough to expose the bug?

smowton reviewed (approved):
Seems fine but the test case is doing lots of stuff that's unrelated to the particular behaviour it's looking to expose

smowton commented:
Test compile failure under Windows was:

```
Failed test: assembly_call_graph_test
file main.c line 12 function thr1: syntax error before `__volatile__'
PARSING ERROR
EXIT=64
SIGNAL=0
```

I guess that's a GCC-ism. Trimming the test case should fix it.

polgreen commented:
The test case is just a copy paste from another regression test; that regression test must have the same gcc error. 

I can minimise but I can't make the gcc-ism go away, because the error is due to the line `__asm__ __volatile__ ("mfence": : :"memory");`, which is the line that makes the reachable call graph fail in develop, and pass with my fix

polgreen commented:
I've minimised the test; it probably won't fix the AWS CodeBuild problem, but I need to keep the assembly in the test, because that is the bug.

smowton commented:
Explanation for the failure not happening otherwise: from `regression/Makefile`:

```
DIRS = cbmc \
       goto-analyzer \
       ansi-c \
       goto-instrument \
       cpp \
       cbmc-cover \
       goto-instrument-typedef \
       smt2_solver \
       strings \
       invariants \
       goto-diff \
       test-script \
       goto-analyzer-taint \
       goto-gcc \
       goto-cc-cbmc \
       cbmc-cpp \
       goto-cc-goto-analyzer \
       systemc \
       contracts
```

i.e. the cbmc-concurrency tests are not run by CI.

smowton commented:
From https://msdn.microsoft.com/en-us/library/5f7adz6y.aspx looks like for Microsoft compilers `__asm mfence` might suffice.

smowton commented on regression/goto-instrument/assembly_call_graph_test/main.c:
```diff
@@ -0,0 +1,13 @@
+void* thr1(void * arg) {
+ __asm__ __volatile__ ("mfence": : :"memory");
```
While you're minimising, might as well kill one of the duplicate `thr` methods

polgreen commented:
The regression test is valid C code,  I would prefer to leave it in the exact format it is in Xen, since that is what I want to compile and run things like slice-global-inits on. 

I don't think regression tests should be required to be C code that compiles on windows, if it is valid C code. Otherwise it's impossible to have regression tests for some of the actual CBMC industrial use cases. 

Is there a way for me to get access to the output from the Aws Code Build please, without having to request it via GitHub?

smowton commented:
Sure, do something like

```
#ifdef __GNUC__
__asm__ __volatile__ ("mfence": : :"memory");
#elif defined(_MSC_VER)
__asm mfence;
#endif
```

That should test the construct in Xen under GCC and compatible compilers, and as an added bonus test the MSVC syntax for the same thing when testing under Win32

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 6fa41f9ccdb12f7f8a941a0b62356a7572f21fbd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80472665
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a73b9d804228cb87f05c7c7dbc4b0640d5da5305).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80473032

polgreen commented:
Changed. 

Perhaps, since 14 of them fail, the CBMC-concurrency tests should be run as part of the regression.

smowton commented:
Not volunteering for the cleanup, but yes

polgreen commented:
Ok, this still fails, can I have the output please?

If it is not possible for external collaborators to get access to AWS Code Build output, I think that either:
- this should be removed as CI and replaced with something external collaborators can see or
- someone that has access to the output should fix the issue in cases where it is something small like this. 

The cycle time for debugging is just too great if I have to keep asking people to copy paste the output to GitHub. 
cc @kroening

polgreen committed:
Make test work on windows

This is a hack, but "__asm__ __volatile__ ("mfence": : :"memory");" doesn't compile on windows and the way CBMC handles "__asm mfence" needs to be fixed. Currently "__asm mfence" is not recognised as a function by CBMC.

polgreen commented:
OK, I got the output from Kroening, CBMC doesn't appear to recognise `__asm mfence` and it doesn't appear in the call graph at all.

That probably needs fixing, but it can't be me that does it. I have replaced the test with a hack so that it should work for windows. 

Ideally I need this fix merged without waiting for someone to fix the way CBMC handles `__asm mfence`, which doesn't sound unreasonable since that is a separate bug.  

Diffblue should also want this PR merged since it fixes a bug which otherwise causes the goto-instrument global init slicer to throw an ugly exception.

peterschrammel commented:
> I can't access the AWS code build

smowton commented on regression/goto-instrument/assembly_call_graph_test/main.c:
```diff
@@ -0,0 +1,20 @@
+void* thr(void * arg) {
+#ifdef __GNUC__
+__asm__ __volatile__ ("mfence": : :"memory");
+#elif defined(_MSC_VER)
```
Make this `#else` in that case

polgreen replied:
this was a straight copy+paste of the code you put above!

smowton replied:
I know, "in that case" was if we're no longer using anything MSVC-specific

peterschrammel commented:
> I can't access the AWS code build

@polgreen, I'll give you access.

smowton commented on regression/goto-instrument/assembly_call_graph_test/main.c:
```diff
@@ -0,0 +1,20 @@
+// windows handles assembly code needs to be fixed.
+// CBMC doesn't recognise __asm mfence as a function.
+
+#ifdef _MSC_VER
```
(and this `#ifndef __GNUC__`)

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 6d0776fb5f7369843904383abe87ed9937120757).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80481544

polgreen commented:
changes done

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ea2d3933e3d858cb2ac5a2e444a2c74ebdad0ad8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80493788
